server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,448 @@
+            .map(DiscoveryNodeRole::roleName)
+            .collect(Collectors.toSet());
+        final String symptom;
+        if (clusterHasBlockedIndex && allUnhealthyNodes.isEmpty()) {
+            // In this case the disk issue has been resolved but the index block has not been automatically removed yet:
+            symptom = String.format(
+                Locale.ROOT,
+                "%d %s blocked and cannot be updated but 0 nodes are currently out of space.",
+                blockedIndices.size(),
+                blockedIndices.size() == 1 ? "index is" : "indices are"
+            );
+        } else {
+            symptom = String.format(
+                Locale.ROOT,
+                "%d node%s with roles: [%s] %s out of disk or running low on disk space.",
+                allUnhealthyNodes.size(),
+                allUnhealthyNodes.size() == 1 ? "" : "s",
+                allRolesOnUnhealthyNodes.stream().sorted().collect(Collectors.joining(", ")),
+                allUnhealthyNodes.size() == 1 ? "is" : "are"
+            );
+        }
+        return symptom;
+    }
+
+    private List<HealthIndicatorImpact> getImpacts(
+        Set<String> indicesWithBlock,
+        Set<String> indicesOnRedNodes,
+        Set<String> indicesOnYellowNodes,
+        Set<String> nodesWithBlockedIndices,
+        Set<String> redDataNodes,
+        Set<String> yellowDataNodes,
+        Set<String> redMasterNodes,
+        Set<String> yellowMasterNodes,
+        Set<String> redNonDataNonMasterNodes,
+        Set<String> yellowNonDataNonMasterNodes
+    ) {
+        List<HealthIndicatorImpact> impacts = new ArrayList<>();
+        if (indicesWithBlock.isEmpty() == false
+            || indicesOnRedNodes.isEmpty() == false
+            || nodesWithBlockedIndices.isEmpty() == false
+            || redDataNodes.isEmpty() == false) {
+            impacts.add(
+                new HealthIndicatorImpact(1, "Cannot insert or update documents in the affected indices.", List.of(ImpactArea.INGEST))
+            );
+        } else if (indicesOnYellowNodes.isEmpty() == false || yellowDataNodes.isEmpty() == false) {
+            impacts.add(
+                new HealthIndicatorImpact(
+                    1,
+                    "At risk of not being able to insert or update documents in the affected indices.",
+                    List.of(ImpactArea.INGEST)
+                )
+            );
+        }
+        if (redMasterNodes.isEmpty() == false || yellowMasterNodes.isEmpty() == false) {
+            impacts.add(new HealthIndicatorImpact(2, "Cluster stability might be impaired.", List.of(ImpactArea.DEPLOYMENT_MANAGEMENT)));
+        }
+        if (redNonDataNonMasterNodes.isEmpty() == false || yellowNonDataNonMasterNodes.isEmpty() == false) {
+            impacts.add(
+                new HealthIndicatorImpact(2, "Some cluster functionality might be unavailable.", List.of(ImpactArea.DEPLOYMENT_MANAGEMENT))
```

Comment:
I think the node roles that are affected should define the functionality that might be impaired (ie. ML, transform, ingest pipelines).

Let's fix this in a subsequent PR (create an issue for this)

Reply:
Created: https://github.com/elastic/elasticsearch/issues/90266